File: ecommerce-go-fiber/apps/orders/internal/config/env.go
package config

import (
	"fmt"
	"os"
	"strconv"
)

type Env struct {
	Port             string
	ServiceName      string
	PostgresHost     string
	PostgresPort     int
	PostgresDB       string
	PostgresUser     string
	PostgresPassword string
	ProductsHTTPURL  string
	ProductsGRPCAddr string
	RabbitURL        string
}

func required(name string) string {
	value := os.Getenv(name)
	if value == "" {
		panic(fmt.Sprintf("missing env var: %s", name))
	}
	return value
}

func getString(name, fallback string) string {
	value := os.Getenv(name)
	if value == "" {
		return fallback
	}
	return value
}

func getInt(name string, fallback int) int {
	raw := os.Getenv(name)
	if raw == "" {
		return fallback
	}

	n, err := strconv.Atoi(raw)
	if err != nil {
		panic(fmt.Sprintf("invalid env var %s=%q", name, raw))
	}

	return n
}

func LoadEnv() Env {
	return Env{
		Port:             getString("PORT", "4020"),
		ServiceName:      getString("SERVICE_NAME", "orders"),
		PostgresHost:     getString("POSTGRES_HOST", "localhost"),
		PostgresPort:     getInt("POSTGRES_PORT", 5432),
		PostgresDB:       getString("POSTGRES_DB", "orders"),
		PostgresUser:     getString("POSTGRES_USER", "admin"),
		PostgresPassword: getString("POSTGRES_PASSWORD", "password"),
		ProductsHTTPURL:  getString("PRODUCTS_HTTP_URL", "http://localhost:4030"),
		ProductsGRPCAddr: getString("PRODUCTS_GRPC_ADDRESS", "0.0.0.0:50052"),
		RabbitURL:        getString("RABBIT_URL", "amqp://admin:password@localhost:5672"),
	}
}

File: ecommerce-ts-express/services/orders/src/config/env.ts
export type Env = {
  port: number
  serviceName: string

  postgresHost: string
  postgresPort: number
  postgresDb: string
  postgresUser: string
  postgresPassword: string
  productsBaseUrl: string
  productsGrpcAddress: string
  rabbitUrl: string
}

function required(name: string, value: string | undefined): string {
  if (!value) throw new Error(`Missing env var: ${name}`)
  return value
}

function toNumber(
  name: string,
  value: string | undefined,
  fallback: number
): number {
  const raw = value ?? String(fallback)
  const n = Number(raw)
  if (!Number.isFinite(n))
    throw new Error(`Invalid number env var: ${name}="${raw}"`)
  return n
}

function toString(
  name: string,
  value: string | undefined,
  fallback: string
): string {
  const raw = value ?? fallback

  if (raw === undefined || raw === null || raw.trim() === '') {
    throw new Error(`Invalid string env var: ${name}`)
  }

  return raw
}

export function getEnv(): Env {
  return {
    port: toNumber('PORT', process.env.PORT, 3020),
    serviceName: process.env.SERVICE_NAME ?? 'orders',

    postgresHost: process.env.POSTGRES_HOST ?? 'localhost',
    postgresPort: toNumber('POSTGRES_PORT', process.env.POSTGRES_PORT, 5432),
    postgresDb: toString('POSTGRES_DB', process.env.POSTGRES_DB, 'orders'),
    postgresUser: toString('POSTGRES_USER', process.env.POSTGRES_USER, 'admin'),
    postgresPassword: toString(
      'POSTGRES_PASSWORD',
      process.env.POSTGRES_PASSWORD,
      'password'
    ),
    productsBaseUrl:
      process.env.PRODUCTS_BASE_URL ?? 'http://localhost:3030/api/products',
    productsGrpcAddress: process.env.PRODUCTS_GRPC_ADDRESS ?? 'localhost:50051',
    rabbitUrl: process.env.RABBIT_URL ?? 'amqp://admin:password@localhost:5672'
  }
}
